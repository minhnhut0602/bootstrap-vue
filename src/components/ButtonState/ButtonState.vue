<template>
    <Button class="">
        <span
            v-if="error || loading || success"
            class="me-2"
        >
            <Spinner
                v-if="loading"
                size="sm"
                variant=""
            />

            <span
                v-if="success"
                class="button-state-icon button-state-icon-error"
            />

            <span
                v-if="error"
                class="button-state-icon button-state-icon-success"
            />
        </span>

        <slot />
    </Button>
</template>

<script lang="ts" setup>
import Button from '../Button';
import Spinner from '../Spinner';

defineProps({
    error: {
        type: Boolean,
        default: false,
    },
    loading: {
        type: Boolean,
        default: false,
    },
    success: {
        type: Boolean,
        default: false,
    },
});
</script>
