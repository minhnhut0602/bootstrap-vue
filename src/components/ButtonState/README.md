# ButtonState

A button that can display certain states wich are usually paired with asynchronous events.

```html
<ButtonState
    :error="error"
    :loading="loading"
    :success="success"
>
    Submit
</ButtonState>
```

## Props

| Name | Type | Description |
| - | - | - |
| error | boolean | Controls the error state of the button. |
| errorVariant | [Variant](#link-to-variant-definition) (string) | The variant of the button in the error state. |
| loading | boolean | Controls the loading state of the button. |
| loadingVariant | [Variant](#link-to-variant-definition) (string) | The variant of the button in the loading state. |
| success | boolean | Controls the success state of the button. |
| successVariant | [Variant](#link-to-variant-definition) (string) | The variant of the button in the success state. |
| variant | [Variant](#link-to-variant-definition) (string) | The variant of the button in idle state. |

## Slots

| Name | Description |
| - | - |
| default | Displays content inside the button. |
