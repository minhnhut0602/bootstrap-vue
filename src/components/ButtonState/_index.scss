.button-state-icon {
    display: inline-block;
    width: $btn-font-size;
    height: $btn-font-size;
    vertical-align: $spinner-vertical-align;
    background-repeat: no-repeat;
    background-position: center;
    background-size: 100%;
}

.button-state-icon-error {
    // TODO: `fill` is not reactive to button styling.
    background-image: escape-svg(url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 8 8'><path fill='currentColor' d='M2.3 6.73L.6 4.53c-.4-1.04.46-1.4 1.1-.8l1.1 1.4 3.4-3.8c.6-.63 1.6-.27 1.2.7l-4 4.6c-.43.5-.8.4-1.1.1z'/></svg>"));
}

.button-state-icon-success {
    // TODO: `fill` is not reactive to button styling.
    background-image: escape-svg(url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 12 12' width='12' height='12' fill='none' stroke='currentColor'><circle cx='6' cy='6' r='4.5'/><path stroke-linejoin='round' d='M5.8 3.6h.4L6 6.5z'/><circle cx='6' cy='8.2' r='.6' fill='currentColor' stroke='none'/></svg>"));
}
