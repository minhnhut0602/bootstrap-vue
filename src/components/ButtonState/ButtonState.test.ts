import ButtonState from '.';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(ButtonState);

    componentSlotRenderTest(ButtonState);
});

describe('disabled', () => {
    it('is false when loading', () => {
        const wrapper = shallowMount(ButtonState, {props: {loading: false}});

        expect(wrapper.vm.disabled).toBe(false);
    });

    it('is true when loading', () => {
        const wrapper = shallowMount(ButtonState, {props: {loading: true}});

        expect(wrapper.vm.disabled).toBe(true);
    });
});

describe('variant', () => {
    it('is variant prop by default', () => {
        const wrapper = shallowMount(ButtonState, {props: {variant: 'info'}});

        expect(wrapper.vm.variant).toBe('info');
    });

    it('is successVariant prop when success', () => {
        const wrapper = shallowMount(ButtonState, {
            props: {
                success: true,
                successVariant: 'info',
            },
        });

        expect(wrapper.vm.variant).toBe('info');
    });

    it('is errorVariant prop when error', () => {
        const wrapper = shallowMount(ButtonState, {
            props: {
                error: true,
                errorVariant: 'info',
            },
        });

        expect(wrapper.vm.variant).toBe('info');
    });
});
