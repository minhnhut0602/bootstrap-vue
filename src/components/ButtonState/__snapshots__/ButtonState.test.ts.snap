// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `"<button-stub disabled=\\"false\\" outline=\\"false\\" size=\\"md\\" variant=\\"primary\\"></button-stub>"`;
